site/content/blog/2021-11-17-database-sql.md

```diff
@@ -0,0 +1,219 @@
+---
+title: Re-implementing the database/sql API on top of RPC
+slug: database-sql
+date: 2021-11-17
+categories:
+    - "go"
+author: Agniva De Sarker
+github: agnivade
+community: agnivade
+---
+
+The `database/sql` package in the Go standard library maintains a pool of connections so that all queries going through a single `*sql.DB` instance will reuse the same pool. This is great because you get a connection pool out of the box. But what if you need to share the same connection pool across processes? How do you use the same API in different processes but still reuse the same pool?
+
+### Background
+
+Mattermost has a plugin architecture to extend its functionality to a user's workflow. These plugins are separate processes running independently from the main Mattermost server. And they typically communicate with the server via an RPC API.
+
+This has served us well for several years, but as plugins started to be more powerful, they needed to maintain their own state in separate database tables. Initially, we just had a single table offering key-value semantics for all plugins via the plugin API, but clearly that wasn't enough. Plugins needed access to the database.
+
+This created a new challenge. Since each plugin was its own isolated process, each of them would open additional connections to the DB. And the more plugins we had, the more amplified this would become. Our customers started to complain that the system exceeded the `max_connections` config setting on their DB.
```

Comment:
```suggestion
This created a new challenge. Since each plugin was its own isolated process, each of them would open additional connections to the DB. And the more plugins we had, the more amplified this would become. Our users started to complain that the system exceeded the `max_connections` config setting on their DB.
```